Add Hellinger Distance for univariate drift detection on continuous data (#149)
NannyML/nannyml: docs/how_it_works/univariate_drift_detection.rst

.. math::
    W_1\left(X_i^{ref},X_i^{ana}\right) = \int_\mathbb{R}\left|\hat{F}_{ref}(x)-\hat{F}_{ana}(x)\right|dx

.. _univariate-drift-detection-cont-hellinger:

Hellinger Distance
........................

The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 

> to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 

Reviewer: It would be good to say that reference and analysis are samples. Also - and this likely applies for other methods as well - I m not too keen on naming them reference and analysis. Because while we may be using the whole reference dataset on one side, we are only using a chunk on the other. Speaking about analysis here may confuse people.

Reviewer 2: Damn, good point. Maybe in how it works where, we describe all the methods, we should just name them p and q, or sample A and B, or whatever. Apart from that we should have a paragraph that explains how this works in the library, that reference set serves as sample A, and every chunk of data that is inputed to _calculate/estimate_ method is the B sample?

analysis assigns some positive probability and vice versa. 
The formula is given by:

.. math::
    H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\int_{}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2dx\right]^{1/2}

> The implementation splits a continuous feature into bins, calculates the relative frequency

Reviewer: IT would be good to improve the language a bit here, for example:
```suggestion
In order to Calculate Hellinger Distance NannyML splits a continuous feature into bins, calculates the relative frequency
```

for each bin from reference and analyzed data and calculates the
resulting Hellinger Distance. The binning is done using `Doane's formula`_ from numpy.
If a continuous feature has relatively low amount of unique values, meaning that
unique values are less then 10% of the reference dataset size up to a maximum of 50, each value becomes a bin.

> This distance is very closely related to the Bhattacharya Coefficient. However we choose the former because it follows the triangle inequality and is 

Reviewer: Since in the implementation we are not using BC I m not sure we want have this in our docs.

Thoughts @jakubnml ?

We see how the relative frequencies of three categories have changed between reference and analysis data.
We also see that the JS Divergence contribution of each change and the resulting JS distance.

.. _univ_cat_method_hellinger:

Hellinger Distance
........................

> The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 

Reviewer: ```suggestion
The Hellinger Distance, is a distance metric used to quantify the similarity between two probability distributions. It measures the overlap between the probabilities assigned 
```

> The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 

Reviewer: Make `Hellinger Distance` a link to the relevant Wiki page?

> to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 

Reviewer: ```suggestion
to the same event by both reference and analysis samples. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 
```

analysis assigns some positive probability and vice versa. 

> The formula is given by:

Reviewer: ```suggestion
```

For a categorical feature Hellinger Distance is defined as:

.. math::

>  H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\sum_{x \in X}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2dx\right]^{1/2}

Reviewer: ```suggestion
 H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\sum_{x \in X}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2\right]^{1/2}
```
Integral ``dx`` leftover from formula here.

> 

Reviewer: It would also be good to explain that F in the categorical case correspond to Probability Mass Functions.
